x/mongo/driver/topology/server_test.go

```diff
@@ -205,30 +193,37 @@ func TestServerConnectionTimeout(t *testing.T) {
 				assert.Nil(t, err, "expected no error but got %s", err)
 			}
 
+			// If we expect the pool to be cleared, watch for all events until we get a
+			// "ConnectionPoolCleared" event or until we hit a 10s time limit.
+			if tc.expectPoolCleared {
+				assert.Eventually(t,
+					func() bool {
+						for evt := range eventsCh {
+							if evt.Type == event.PoolCleared {
+								return true
+							}
+						}
+						return false
+					},
+					10*time.Second,
+					100*time.Millisecond,
+					"expected pool to be cleared within 10s but was not cleared")
+			}
+
 			// Disconnect the server then close the events channel and expect that no more events
-			// are sent on the channel. Then wait for the events channel loop to return before
-			// inspecting the events slice.
+			// are sent on the channel.
 			_ = server.Disconnect(context.Background())
-
 			close(eventsCh)
```

Comment:
Am I correct that this test may have failed before if `close(eventsCh)` ran before the "ConnectionPoolCleared" event was emitted? If so, nice find. The new approach seems more resilient.

Reply:
`pool.close()` (called by `server.Disconnect()`) waits for all background goroutines to exit before returning, so it shouldn't be possible for events to be sent after `server.Disconnect()` returns. The problem is actually that `pool.clear()` was called after `pool.close()` in the "operation context timeout with unrelated dial errors should clear the pool" test case. After responding to @benjirewis's feedback, I updated `pool.clear()` to be a no-op if the pool is "closed", so the "ConnectionPoolCleared" event was no longer sent and the test failed. The updated test waits for the "ConnectionPoolCleared" event to be sent before calling `server.Disconnect()`, resolving that issue.

Reply:
Thank you for the explanation.